Add Env variables in OTLP exporter

jan25 opened the pull request:
# Description

Adds support for Env variables in OTLP trace and metrics exporter. Implemented subset of variables from [exporter spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/protocol/exporter.md#configuration-options), other variables will be implemented as the usecases are added to otlp exporter.

Fixes #1004 (issue)

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] Write unit tests

# Checklist:

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

owais commented on exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/exporter.py:
```diff
@@ -112,18 +119,22 @@ class OTLPExporterMixin(ABC):
         super().__init__()
 
+        endpoint = endpoint or os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT") or "localhost:55680"
+        insecure = insecure or os.environ.get("OTEL_EXPORTER_OTLP_INSECURE") or True # TODO(jan25): verify spec for default value. This affects how credentials are passed
```
Does this mean values like `"false"`, `"no"`, `"off"` and their upper-case versions will evaluate to true?

owais replied:
I think we should use `opentelemetry.configuration.Configuration` class to read these env vars. I think automatically casts some values to Python types and is future proof in the sense that any changes to how we prefix env vars or whether we support config files in future will not warrant a change here.

jan25 replied:
Thanks for the hint. I adjusted the implementation to use `Configuration` object.

owais commented on exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/exporter.py:
```diff
@@ -112,18 +119,22 @@ class OTLPExporterMixin(ABC):
+        if insecure:
             self._client = self._stub(insecure_channel(endpoint))
         else:
+            credentials = credentials or _load_credential_from_file(os.environ.get("OTEL_EXPORTER_OTLP_CERTIFICATE"))
```
If the file fails to load, what error message will the user see? Will it be helpful/obvious enough? Perhaps we shoudl wrap the exception with a more helpful and obvious error message. Something like, "Failed to load certificate from <path> as referenced by OTEL_EXPORTER_OTLP_CERTIFICATE"

jan25 replied:
I've wrapped the file reading in a try block. Now it logs and returns `None` in case of exception.

owais commented on exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/metrics_exporter/__init__.py:
```diff
@@ -138,6 +140,26 @@ class OTLPMetricsExporter(MetricsExporter, OTLPExporterMixin):
+        insecure: bool,
+        credentials: ChannelCredentials = None,
+        metadata: tuple = None,
+    ):
```
same concerns as above

jan25 commented:
We need to wait for [this PR for spec](https://github.com/open-telemetry/opentelemetry-specification/pull/975) updates, so we follow the correct default for `OTEL_.._INSECURE` env variable.

Also, for PR checklist:
* I'll update readme file for exporter to mention about env variables
* Unit tests could be tricky to implement, my idea is to start a grpc server in unit test with tls enabled and listening to specific endpoint/port and client sends a sample span to the server through using env vars. Any suggestions here?

lzchen commented on exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/exporter.py:
```diff
@@ -127,18 +141,33 @@ class OTLPExporterMixin(
+        if insecure is None:
+            insecure = Configuration().EXPORTER_OTLP_INSECURE
+        if insecure is None:
+            insecure = True
```
Isn't the default value `False`?

jan25 replied:
In current spec it is `False`. I [opened a proposal PR](https://github.com/open-telemetry/opentelemetry-specification/pull/975) for changing default to be `True` which feels more natural to have, otherwise supplying certificate files becomes mandatory.

lzchen replied:
I would prefer the implementation to follow the specs AS OF NOW. If the PR in specs ever gets merged, then we can create a followup PR to change that. That way we avoid confusion (like how I was confused just now).

lzchen commented:
@jan25 
Can you add an entry in the CHANGELOG?

>Unit tests could be tricky to implement, my idea is to start a grpc server in unit test with tls enabled and listening to specific endpoint/port and client sends a sample span to the server through using env vars. Any suggestions here?

This seems like an integration test. For a unit test, you simply need to test the constructor populates the desired fields for each use case, as well as your file opening logic works as expected for credentials.

codeboten commented:
@jan25 is this still a WIP?

jan25 commented:
@codeboten yes, I'll add unit tests and unmark WIP.

lzchen commented on exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/exporter.py:
```diff
@@ -121,24 +134,40 @@ class OTLPExporterMixin(
+        if insecure is None:
+            insecure = False
+
+        self._metadata = metadata or Configuration().EXPORTER_OTLP_HEADERS
```
Should this be called `self._headers`?

jan25 replied:
This was introduced probably because [grpc client arg](https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-proto/src/opentelemetry/proto/collector/trace/v1/trace_service_pb2_grpc.py#L70) is named metadata too. Do you suggest we rename on exporter side to headers?

lzchen replied:
Yes. It matches what the env variable name is.

lzchen commented on exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/exporter.py:
```diff
@@ -121,24 +134,40 @@ class OTLPExporterMixin(
         self,
-        endpoint: str = "localhost:55680",
+        endpoint: Optional[str] = None,
+        insecure: Optional[bool] = None,
```
Missing `protocol`, `certificate_file` in constructor, `compression` and `timeout`?

jan25 replied:
We could add env vars for `protocol` `compression` when those usecases are implemented. Currently exporter uses `grpc` only and no implementation for `compression`.

`certificate_file` is wrapped in ChannelCredentials object which is passed down from `OTLPSpanExporter|MetricExporter`. I'll change `timeout` logic to use env var too.

codeboten replied:
Compression is implemented in a separate PR https://github.com/open-telemetry/opentelemetry-python/pull/1141/files#diff-77db88bbe988c1d7f7d9f92f9e05400259a026e7203f7e9ee12aef76b3b0fdb3R142

lzchen commented on exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/exporter.py:
```diff
@@ -121,24 +134,40 @@ class OTLPExporterMixin(
+            or "localhost:55680"
+        )
+
+        if insecure is None:
```
Nit: Why is the logic not the as for `endpoint`?

jan25 replied:
I think we want to have this precedence: `insecure arg` > `Trace/Metric exporter specific env var` > `Generic env var`. Because `insecure` is a boolean it'd be tricky to use chained `or`.

lzchen reviewed (changes requested):
Some comments.

codeboten commented:
@jan25 looks like some tests are still failing, is the PR still a work in progress or is it ready for review?

jan25 commented:
@codeboten Tests are failing on older Python versions due to unittest api changes, I've tried a few things but didn't succeed, I'll have to fix them. Otherwise this is ready to review

codeboten reviewed (changes requested):
Overall this looks pretty good! Just a request for change on the changelog. It would be good have information about the exporter environment variables in docs somewhere

jan25 commented:
@codeboten Thanks! I've fixed the tests and updated changelog too.

Do you suggest we add a link to [exporter-spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/protocol/exporter.md#opentelemetry-protocol-exporter) in `README.rst` file? Perhaps we can also add about whats currently implemented?

codeboten commented:
> Do you suggest we add a link to [exporter-spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/protocol/exporter.md#opentelemetry-protocol-exporter) in `README.rst` file? Perhaps we can also add about whats currently implemented?

I added a note to capture this work in the issue to review documentation, I don't want to block this PR on it and there are a bunch of places we need to review env variables anyways https://github.com/open-telemetry/opentelemetry-python/issues/1133#issuecomment-718081899

codeboten reviewed (approved):
PR looks great!

NathanielRN commented on exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/exporter.py:
```diff
@@ -113,6 +116,16 @@ def _get_resource_data(
 
+def _load_credential_from_file(filepath) -> ChannelCredentials:
+    try:
+        with open(filepath, "rb") as f:
```
Tried to use the OTLP Exporter today and I think I ran into an issue. If you don't have the `OTEL_PYTHON_EXPORTER_OTLP_CERTIFICATE` Env variable set, then the call to `Configuration().EXPORTER_OTLP_CERTIFICATE` 

https://github.com/jan25/opentelemetry-python/blob/2ff2492333ab85d10b81a89f86d48a6eaad74912/exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/exporter.py#L175

will return `None` instead of a `String`. Then this line will throw error 

`TypeError: expected str, bytes or os.PathLike object, not NoneType`

which is not caught by `FileNotFoundError` below.

lzchen replied:
I suggest opening a new issue to track these problems instead of continuing an already closed PR thread.

NathanielRN commented on exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/trace_exporter/__init__.py:
```diff
@@ -50,13 +55,47 @@ class OTLPSpanExporter(
+                "endpoint": endpoint
+                or Configuration().EXPORTER_OTLP_SPAN_ENDPOINT,
+                "insecure": insecure,
+                "credentials": credentials,
```
If I set `OTLPExporter(insecure=True)` this works fine, but if I leave that out, I get a message

```
File "/Users/enowell/git/opentelemetry-python/exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/exporter.py", line 178, in __init__
    self._client = self._stub(secure_channel(endpoint, credentials))
  File "/Users/enowell/Library/Python/3.8/lib/python/site-packages/grpc/__init__.py", line 1943, in secure_channel
    if credentials._credentials is _insecure_channel_credentials:
AttributeError: 'NoneType' object has no attribute '_credentials'
```

which makes sense because `insecure=None` and `credentials=None`. I think a follow up would be to throw an Error Message if `insecure=None (or False) && credentials=None` saying "Error: Either provide credentials or label as insecure."

rgstephens replied:
@NathanielRN Thanks for reporting this. I also ran into this issue. Causes a failure in the example code.

NathanielRN reviewed (commented):
This works, just leaving my comments about some UX issues I had today.

jan25 commented:
@NathanielRN thanks for the review! I'll work on fixing them and open a PR. I couldn't think of way to add such case in unittests, will try again.
